readme.md
# Running the Housing Market model

## Setup

### First Time
To install the environment, install pipenv onto your computer.<br>
For first time users, first run pipenv install from your shell in the right repository.<br>
Warning, this might take a while!

### Normal setup
From then on out, you can run 'pipenv shell' in the repository, followed by 'ipyhton server.py'.<br>
This will open a window on your default browser in which you can see the model.

Adjust the speed at which the model runs with the slider in the top right.

The squares represent houses, while dots represent households. A house will be green when avalaible, <br>
red when unavalaible. 

## The Model

### Basic Ideas
This model aims to model the dutch housing market. It shows how inequality increases in a simple representation of a market without<br>
inheritance and people owning multiple homes. It draws agent attributes from real data from the dutch market.<br>
Agents are bounded rational, and try to maximize their profits based on their predictions of the market. 


### Altering the model
The model can be toyed with by changing it's default values in model.py. Furthermore, one could try to add things to the model

changes.txt
1. Initiliaze population:
    - Uniform age distribution between 20 and 99
    - Draw income from age distribution, so it's not a paramater
    - Housing and rental prices from CBS distribution
    - Initial savings from normal distribution mean 50000, variance 50000
    - Risk averse index
2. Burnin period:
    - Let the model run for 80 steps, now there is a whole new population
    - Start measuring the effect of initial savings for whealth throughout life
3. During the run:
    - When an agent dies, initialize a new one
    - When agent ages, change there income
    - At every step, give them a chance to list their home based on utility

- Research question:
    How does an initial saving influence your wealth at age 70?

    How does initials savings incfluence the Gini coefficient?

    -Subquestion:
        What risk strategy is best when your poor/rich? 


model.py
import random
import numpy as np
import csv
#
from mesa import Model
from mesa.space import MultiGrid
from mesa.datacollection import DataCollector
from mesa.time import RandomActivation, StagedActivation
from agents import *
from datacollection import *


class HouseActivation(RandomActivation):
    def __init__(self, model):
        super().__init__(model)
        self.model = model

    def get_available(self):
        return [house for house in self.model.schedule_House.agents if house.available]


class HousingMarket(Model):
    def __init__(self, height=20, width=20, initial_houses=100, initial_households=150,
                 savings_lower=0, savings_upper=100000, price_lower=100000, price_upper=1000000,
                 payoff_perc_freehold=0.0025, inflation=0.02, house_price=400_000,
                 chi_parameter=6.5, maximum_age=100, minimum_age=20, age_utility_scaling = 0.01,
                 maximum_moving_age=65, bank_income_multiplier=8, fraction_good_houses=0.5,
                 price_shock_range=6):
        super().__init__()
        self.height = width
        self.width = height
        self.initial_houses = initial_houses
        self.initial_households = initial_households
        self.rentals = self.initial_households - initial_houses
        self.house_price = house_price
        self.chi_parameter = chi_parameter
        self.maximum_age = maximum_age
        self.minimum_age = minimum_age
        self.age_utility_scaling = age_utility_scaling
        self.maximum_moving_age = maximum_moving_age
        self.bank_income_multiplier = bank_income_multiplier
        self.fraction_good_houses = fraction_good_houses
        self.price_shock_range = price_shock_range

        # determines the percentage of house price that you pay off
        self.payoff_perc_freehold = payoff_perc_freehold

        self.savings_lower = savings_lower
        self.savings_upper = savings_upper
        self.price_lower = price_lower
        self.price_upper = price_upper
        self.incomes, self.income_distr = self.draw_income_distribution()
        self.ages, self.age_distr = self.draw_age_distribution()

        self.inflation = inflation
        self.total_inflation = 0
        self.yearly_inflation = 0
        self.income_distribution = np.load("Income Data/income_distribution.npy")

        self.grid = MultiGrid(self.width, self.height, torus=True)

        self.schedule_House = HouseActivation(self)
        self.schedule_Household = RandomActivation(self)
        self.schedule = HouseActivation(self)
        self.running = True

        self.n_households = self.initial_households

        # keep track of number of periods that the model goes through, one step increases the period by 1
        self.period = 0

        self.datacollector = DataCollector(
            model_reporters={
                "Gini": gini_coefficient,
                "Average Savings": average_savings,
                "Age -25 Savings": age_25_minus_savings,
                "Age 25-34 Savings": age_25_34_savings,
                "Age 35-44 Savings": age_35_44_savings,
                "Age 45-54 Savings": age_45_54_savings,
                "Age 55-64 Savings": age_55_64_savings,
                "Age 65-74 Savings": age_65_74_savings,
                "Age 75+ Savings": age_75_plus_savings,
                "Average Household Income": average_household_income,
                'Mean Household Age': mean_household_age,
                'Mean House Price': mean_house_price,
                'Mean Forecasted House Price Change': mean_house_price_change,
                "Age -25 amount": age_25_amount,
                "Age 25-34 amount": age_25_34_amount,
                "Age 35-44 amount": age_35_44_amount,
                "Age 45-54 amount": age_45_54_amount,
                "Age 55-64 amount": age_55_64_amount,
                "Age 65-74 amount": age_65_74_amount,
                "Age 75+ amount": age_75_plus_amount,
                "Agent count": total_agents,
                "Inflation": get_inflation,
                "Total Inflation": get_total_inflation
            },
            agent_reporters={
                "Income": collect_income,
                "Price": 'price',
                "Age": collect_ages,
            })

        self.initialize_population(House, self.initial_houses)
        self.initialize_population(Household, self.initial_households)
        self.assign_houses()

    def draw_income_distribution(self):
        incomes = []
        counts = []

        with open('input_data/incomes.csv') as file:
            csv_reader = csv.reader(file, delimiter=',')
            for row in csv_reader:
                incomes.append([int(row[0]) * 1000 / 12, int(row[1]) * 1000 / 12])
                counts.append(int(row[2]))

        cum_counts = []
        for i in range(len(counts)):
            cum_counts.append(sum(counts[:i + 1]))

        hhld_count = sum(counts)

        cum_ratios = [x / hhld_count for x in cum_counts]

        return incomes, cum_ratios

    def draw_age_distribution(self):
        ages = []
        age_counts = []

        with open('input_data/ages.csv') as file:
            csv_reader = csv.reader(file, delimiter=',')
            for row in csv_reader:
                ages.append(int(row[0]))
                age_counts.append(int(row[1]))

        # we consider housing market participants in age range 20-100
        ages = ages[20:]
        age_counts = age_counts[20:]

        '''
        cum_counts = []
        for i in range(len(age_counts)):
            cum_counts.append(sum(age_counts[:i + 1]))

        pop_count = sum(age_counts)

        cum_ratios = [x / pop_count for x in cum_counts]
        '''

        return ages, [ages, age_counts]

    def initialize_population(self, agent_type, n):
        for i in range(n):
            x = random.randrange(self.width)
            if x == 0:
                y = random.randrange(1, self.height)
            else:
                y = random.randrange(self.height)

            self.new_agent(agent_type, (x, y))

    def assign_houses(self):
        _len = len(self.schedule_House.agents)
        if len(self.schedule_Household.agents) < len(self.schedule_House.agents): _len = len(
            self.schedule_Household.agents)
        for i in range(_len):
            self.schedule_House.agents[i].set_availability(False)
            self.schedule_Household.agents[i].house = self.schedule_House.agents[i]
            self.schedule_House.agents[i].owner = self.schedule_Household.agents[i]
            MultiGrid.move_agent(self=self.grid, agent=self.schedule_Household.agents[i],
                                 pos=self.schedule_House.agents[i].pos)
    '''
    def init_population(self, agent_type, n):
        for i in range(n):
            x = random.randrange(self.width)
            y = random.randrange(self.height)

            self.new_agent(agent_type, (x, y))
    '''         

    def new_agent(self, agent_type, pos):
        '''
        Method that creates a new agent, and adds it to the correct scheduler.
        '''
        if isinstance(agent_type, Household):
            self.n_households += 1

        agent = agent_type(self.next_id(), self, pos)

        self.grid.place_agent(agent, pos)
        getattr(self, f'schedule_{agent_type.__name__}').add(agent)
        getattr(self, "schedule").add(agent)

    def remove_agent(self, agent):
        '''
        Method that removes an agent from the grid and the correct scheduler.
        '''
        if isinstance(agent, Household):
            self.n_households -= 1

        self.grid.remove_agent(agent)
        getattr(self, f'schedule_{type(agent).__name__}').remove(agent)
        getattr(self, "schedule").remove(agent)

    def step(self):
        '''
        Method that calls the step method
        '''
        #print(f"Step: {self.schedule.steps}") 

        self.schedule.steps += 1

        """ Calculate monthly adjusted inflation and adjust incomes based on that """
        # Derived from Historical CPI data (US 2010->2021)
        self.inflation = np.random.normal(loc=.00186, scale=.00115, size=1)[0]
        self.total_inflation += self.inflation
        self.income_distribution[:, 1] *= 1 + self.inflation
        self.yearly_inflation += self.inflation

        # Introduce a market shock every year
        if self.period % 12 == 0:
            self.house_price_shock = random.uniform(-0.5*self.price_shock_range + 100*self.yearly_inflation,0.5*self.price_shock_range + 100*self.yearly_inflation)
            self.yearly_inflation = 0
    

        self.schedule_House.step()

        self.schedule_Household.step()
        self.datacollector.collect(self)

        # check if population is still of same size 
        self.n_households = len(self.schedule_Household.agents)
        if self.n_households != self.initial_households:
            # if someone has died, add a new household agent to the model 
            n_deaths = self.initial_households - self.n_households
            self.initialize_population(Household, n_deaths)

        self.period += 1

    def run_model(self, step_count=2):
        '''
        Method that runs the model for a specific amount of steps.
        '''
        for i in range(step_count):
            self.step()


server.py
from mesa.visualization.modules import CanvasGrid
from mesa.visualization.ModularVisualization import ModularServer
from mesa.visualization.modules import ChartModule
from mesa.visualization.ModularVisualization import VisualizationElement
from model import *

import IPython
import os
import sys
import numpy as np

# Change stdout so we can ignore most prints etc.
orig_stdout = sys.stdout
sys.stdout = open(os.devnull, 'w')
sys.stdout = orig_stdout


def choose_color(agent):
    if type(agent) is House:
        if agent.available: color = 'green'
        else:   color = 'red'
    else:   color = 'black'
    return color


def agent_portrayal(agent):
    if type(agent) is House:
        portrayal = {"Shape": "rect",
                     "Color": choose_color(agent),
                     "Filled": "true",
                     "Layer": 0,
                     'w': 0.99,
                     'h': 0.99}
    else:
        portrayal = {"Shape": "circle",
                     "Color": choose_color(agent),
                     "Filled": "true",
                     "Layer": 1,
                     'r': .4}
    return portrayal


grid = CanvasGrid(agent_portrayal, 20, 20, 500, 500)

""" Create a dynamic linegraph """
gini_char = ChartModule(
                    [{"Label": "Gini", "Color": "red"}],
                    data_collector_name='datacollector',
                    canvas_width=500,
                    canvas_height=200)


average_savings_chart = ChartModule([
                    {"Label": "Average Savings", "Color": 'red'},
                    {"Label": "Age -25 Savings", "Color": "green"},
                    {"Label": "Age 25-34 Savings", "Color": "blue"},
                    {"Label": "Age 35-44 Savings", "Color": "pink"},
                    {"Label": "Age 45-54 Savings", "Color": "purple"},
                    {"Label": "Age 55-64 Savings", "Color": "black"},
                    {"Label": "Age 65-74 Savings", "Color": "yellow"},
                    {"Label": "Age 75+ Savings", "Color": "orange"}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

average_income_chart = ChartModule([
                    {"Label": "Average Household Income", "Color": 'red'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')


mean_household_age_chart = ChartModule([
                    {"Label": "Age -25 amount", "Color": "green"},
                    {"Label": "Age 25-34 amount", "Color": "blue"},
                    {"Label": "Age 35-44 amount", "Color": "pink"},
                    {"Label": "Age 45-54 amount", "Color": "purple"},
                    {"Label": "Age 55-64 amount", "Color": "black"},
                    {"Label": "Age 65-74 amount", "Color": "yellow"},
                    {"Label": "Age 75+ amount", "Color": "orange"},
                    {"Label": "Mean Household Age", "Color": 'green'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

mean_house_price_chart = ChartModule([
                    {"Label": "Mean House Price", "Color": 'green'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

mean_forecasted_price_chart = ChartModule([
                    {"Label": "Mean Forecasted House Price Change", "Color": 'green'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

inflation_chart = ChartModule([
                    {"Label": "Inflation", "Color": 'blue'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

total_inflation_chart = ChartModule([
                    {"Label": "Total Inflation", "Color": 'blue'}],
                    canvas_width=500, canvas_height=200,
                    data_collector_name='datacollector')

# Create the server, and pass the grid and the graph
server = ModularServer(
                    HousingMarket,
                    [grid, gini_char, average_savings_chart, average_income_chart, mean_household_age_chart,
                     mean_house_price_chart, mean_forecasted_price_chart, inflation_chart, total_inflation_chart],
                    "Housing Market Model",
                    {'initial_houses': 900,
                     'initial_households': 1000
                     })

server.port = 8522
server.launch()


agents.py
# from zoneinfo import available_timezones
from mesa import Agent
from mesa import Model
from mesa.space import MultiGrid
import random
import scipy
import numpy as np


class Household(Agent):
    def __init__(self, unique_id, model, pos):
        super().__init__(unique_id, model)
        self.pos = pos
        self.model = model
        self.savings = random.randint(self.model.savings_lower, self.model.savings_upper)
        self.age = self.set_age()

        self.income, self.bin, self.percentile = self.initialize_income_bin_percentile()
        self.house = None

        # No one starts with a mortgage
        self.mortgage = 0

        self.monthly_ageing = 0
        self.strategy = np.random.choice(["naive", "sophisticated"])

        # Fix risk attitude parameters
        self.alpha = np.random.normal(loc = 0.79, scale=0.3)
        self.beta = np.random.normal(loc = 1.13, scale=0.66)
        self.lmbda = np.random.normal(loc = 1.35, scale=2.59)

    
    def get_mortgage_quote(self):
        """
        Calculates mortage quote based on monthly household disposable income
        Example:
        - Disposable income (mean) of 3500 per month ~ 42k per year ~ 60k gross per year
        - Based on several mortgage calculators, mortgage is ~ 5-6x gross income. 
        - i.e. 60k should result in 300k. Thus, disposable income is ~ 8x to mortgage
        - (this is a naive and deterministic mortgage quote)
        """
        deterministic_quote = self.income * 12 * self.model.bank_income_multiplier
        return deterministic_quote

    def set_age(self):
        # if model is past initialisation, new agents in the model are "born" at youngest available age

        if self.model.period > 0:
            return self.model.minimum_age

        # if model is initialised, distribute age following Dutch age distribution among agents
        # This is done using Acceptance-Rejection Sampling
        ages = self.model.age_distr[0]
        ages_counts = self.model.age_distr[1]

        for i in range(len(self.model.ages)):
            while True:
                x = random.randrange(0, 80)
                y = random.uniform(0, 250000)

                if ages_counts[x] >= y:
                    # Shift to match age
                    return x + 20

    def initialize_income_bin_percentile(self):
        # Get column relative to the age
        column = 8
        for inx, column_age in enumerate([25, 35, 45, 55, 65, 75]):
            if self.age < column_age:
                column = inx + 2
                break

        rn = np.random.uniform(0, 1, 1)[0]
        for row in self.model.income_distribution:
            if rn < row[column]:
                return row[1], int(row[0]), row[column]


    def step(self):
        """
        Step of an agent represents the actions of the agent during one month
        """
        self.monthly_ageing += 1

        if self.monthly_ageing == 12:
            self.age += 1
            self.monthly_ageing = 0

        # Update the income of the agent
        # print(f" income = {self.income} \n bin = {self.bin} \n percentile = {self.percentile}")
        self.income, self.bin, self.percentile = self.update_income_bin_percentile()
        # calculate equity
        if self.house:
            # print(self.house.price)
            self.savings += self.model.payoff_perc_freehold * self.house.price
            self.equity = self.house.price + self.savings - self.mortgage
        else:
            self.equity = self.savings - self.mortgage

        # all available houses
        available_houses = self.model.schedule_House.get_available()

        # decide whether to sell your house
        if (self.age < self.model.minimum_age or self.age > self.model.maximum_moving_age):
            pass

        # depending on their strategy an agent will have a different procedure for deciding whether to sell:
        elif (self.house and self.strategy == "naive"):
            # not everybody is actively checking the market at every step

            if random.random() < (1 - self.model.age_utility_scaling * self.age) or self.empty_neighborhood == True:

                # get new mortgage quote based on current income
                mortgage_quote = self.get_mortgage_quote()

                if self.house:
                    # calculate expected money gained from selling current house
                    house_mortgage_differential = self.house.priceChangeForecast - self.mortgage
                else:
                    house_mortgage_differential = 0

                # calculate total available money for buying a house
                available_money = mortgage_quote + house_mortgage_differential + self.savings

                # sample houses (in neighborhood?)
                house_sample = random.sample(available_houses, k = len(available_houses)) # k is to be adjusted depending on what's realistic
                
                # obtain probability of ending up in a given house:
                attractive_houses = 0
                affordable_houses = 0
                for house in house_sample:
                    if self.house.priceChangeForecast > house.priceChangeForecast and house.price < available_money:
                        attractive_houses += 1
                    if house.price < available_money:
                        affordable_houses += 1

                if attractive_houses == 0 or affordable_houses == 0:
                    prob_buy = 0
                else:
                    prob_buy = attractive_houses/affordable_houses

                # obtain expected utility of buying a new house on the market:
                expected_utility = 0
                for house in house_sample:
                    expected_utility += self.utility(x = house.priceChangeForecast-self.house.priceChangeForecast, alpha = self.alpha, beta = self.beta, lmbda = self.lmbda)*prob_buy
                
                # list own house
                if expected_utility > 0:
                    self.house.set_availability(True)

        elif (self.house and self.strategy == "sophisticated"):
            # not everybody is actively checking the market at every step
            if random.random() < (1 - self.model.age_utility_scaling * self.age) or self.empty_neighborhood == True:

                # get new mortgage quote based on current income
                mortgage_quote = self.get_mortgage_quote()

                if self.house:
                    # calculate expected money gained from selling current house
                    house_mortgage_differential = self.house.priceChangeForecast - self.mortgage
                else:
                    house_mortgage_differential = 0
                    
                # calculate total available money for buying a house
                available_money = mortgage_quote + house_mortgage_differential + self.savings

                # sample houses (in neighborhood?)
                house_sample = random.sample(available_houses, k = len(available_houses))
                
                # obtain probability of ending up in a given house:
                attractive_houses = 0
                affordable_houses = 0
                for house in house_sample:
                    if self.house.priceChangeForecast_av > house.priceChangeForecast_av and house.price < available_money:
                        attractive_houses += 1
                    if house.price < available_money:
                        affordable_houses += 1

                if attractive_houses == 0 or affordable_houses == 0:
                    prob_buy = 0
                else:
                    prob_buy = attractive_houses/affordable_houses

                # obtain expected utility of buying a new house on the market:
                expected_utility = 0
                for house in house_sample:
                    expected_utility += self.utility(x = house.priceChangeForecast_av-self.house.priceChangeForecast_av, alpha = self.alpha, beta = self.beta, lmbda = self.lmbda)*prob_buy
                
                # list own house
                if expected_utility > 0:
                    self.house.set_availability(True)

        # always buy a house if you are renting, this could be enhanced if there was a bidding stage
        elif self.house is None:
            self.buy_house(available_houses)

        # for now implement simple death rule, agent exits model at age of 100
        if self.age == self.model.maximum_age:
            if self.house:
                self.house.set_availability(True)
                self.house.owner = None
            self.model.remove_agent(self)

        # Death dynamics modeled after Gompertz law 
        if self.monthly_ageing == 11:
            if 0.0005 + 10 ** (-4.2 + 0.038 * self.age) >= random.uniform(0, 1):
                if self.house:
                    self.house.set_availability(True)
                    self.house.owner = None
                self.model.remove_agent(self)


    def utility(self, x, alpha, beta, lmbda):
        # This function defines the agents' utility, where x is the expected gain or loss, alpha and beta are risk attitude parameters for gains and losses respectively and lambda is the loss aversion constant.
        if x > 0:
            return x**alpha
        else:
            return (abs(x)**(beta)*lmbda*(-1))
    
    def buy_house(self, available_houses):
        """Method that let's household buy a house from antoher household

        Args:
            available_houses (list): A list of all available houses

        Note: you enter this function with assumption that you do NOT have a house anymore! (otherwise have to change this function)
        """
        # try to buy a house
        if self.strategy == "naive":
            available_houses.sort(key=lambda x: x.priceChangeForecast, reverse=True)
        else:
            available_houses.sort(key=lambda x: x.priceChangeForecast_av, reverse=True)

        for house in available_houses:
            if house.owner == self:
                continue
            # buy the best house avalaible

            mortgage_quote = self.get_mortgage_quote()
            available_money = self.savings + mortgage_quote
            if house.price < available_money:
                # wire the money
                previous_owner = house.owner
                if previous_owner:
                    previous_owner.house = None
                    MultiGrid.move_agent(self=self.model.grid, agent=previous_owner, pos=(0, 0))

                    # pay off mortgage of previous owner and push cash remainder into savings
                    earnings_from_sale = house.price - previous_owner.mortgage
                    previous_owner.savings += earnings_from_sale
                    previous_owner.mortgage = 0

                if house.price > mortgage_quote:
                    # take the mortgage
                    self.mortgage = mortgage_quote

                    # pay remainder with savings
                    pay_with_savings = house.price - mortgage_quote
                    self.savings -= pay_with_savings
                else:
                    # if able to get a mortgage larger than house price
                    # only get mortgage up to house price, and afford entire house with mortgage
                    self.mortgage = house.price

                # change ownership
                self.house = house
                self.house.owner = self
                house.set_availability(False)
                MultiGrid.move_agent(self=self.model.grid, agent=self, pos=house.pos)
                break


    def update_income_bin_percentile(self):
        # Get column relative to the age
        column = 8
        for inx, column_age in enumerate([25, 35, 45, 55, 65, 75]):
            if self.age < column_age:
                column = inx + 2
                break
        random_walk_bin = np.random.normal(loc=0, scale=1, size=1).astype(int)[0]
        for row in self.model.income_distribution:
            if self.percentile <= row[column]:
                bin = int(row[0] + random_walk_bin)
                if bin > 72: bin = 72
                elif bin < 0: bin = 0
                return self.model.income_distribution[bin, 1], bin, self.model.income_distribution[bin, column]
        return self.model.income_distribution[72, 1], 72, self.model.income_distribution[72, column]


    def empty_neighborhood(self):
       # an agent looks around and checks if the majority of houses in their vincinity are empty
       # if they are, the agent decides to move away too
       houses = 0
       for neighbor in self.model.grid.neighbor_iter(self.pos):
           if neighbor.type != self.type:
               houses += 1
       empty_houses = 0
       for neighbor in self.model.grid.neighbor_iter(self.pos):
           if neighbor.type != self.type & neighbor.available:
               empty_houses += 1
       if empty_houses / houses > 0.4:
           return True
       else:
           return False


    def get_mortgage_quote(self):
        """
        Calculates mortage quote based on monthly household disposable income
        Example:
        - Disposable income (mean) of 3500 per month ~ 42k per year ~ 60k gross per year
        - Based on several mortgage calculators, mortgage is ~ 5-6x gross income. 
        - i.e. 60k should result in 300k. Thus, disposable income is ~ 8x to mortgage
        - (this is a naive and deterministic mortgage quote)
        """
        deterministic_quote = self.income * 12 * self.model.bank_income_multiplier
        return deterministic_quote

class House(Agent):
    def __init__(self, unique_id, model, pos):
        super().__init__(unique_id, model)
        self.pos = pos
        # set initial house price
        self.house_price_change = random.random() if random.random() < self.model.fraction_good_houses else random.random() * (-1)
        self.price = self.set_initial_house_price()
        self.priceChange = self.price * random.normalvariate(mu=self.house_price_change,
                                                             sigma=2 * self.house_price_change) / 100
        self.priceChange_past = self.price * random.normalvariate(mu=self.house_price_change,
                                                                  sigma=2 * self.house_price_change) / 100
        self.priceChange_av = (self.priceChange + self.priceChange_past) / 2
        self.owner = None
        self.available = True

    def set_availability(self, set_to):
        self.available = set_to

    def step(self):
        # Price shock once every year
        if self.model.period % 12 == 0:
            self.price += (self.price * random.normalvariate(mu=self.model.house_price_shock,
                                                             sigma=2 * self.model.house_price_shock) / 100)
            # redraw the house change in the new market to introduce some stochasticity
            self.house_price_change = random.random() if random.random() < 0.5 else random.random() * (-1)

        self.priceChange = self.price * random.normalvariate(mu=self.house_price_change,
                                                             sigma=2 * self.house_price_change) / 100
        self.price += self.priceChange
        self.priceChangeForecast = self.price + self.priceChange

        self.priceChange_past = self.priceChange_past + self.priceChange
        self.priceChangeForecast_av = (self.priceChange_past) / (self.model.period + 1)

    def set_initial_house_price(self):
        cd = scipy.random.chisquare(self.model.chi_parameter, size=1)

        """ Scale for Std """
        cd = cd / (2 * self.model.chi_parameter) ** 1 / 2

        """ Adjust Mean so ~= 3484 (mean monthly Dutch Household Income) """
        mean_chi = self.model.chi_parameter / (2 * self.model.chi_parameter) ** 1 / 2
        cd = cd * (1 / mean_chi) * self.model.house_price
        return cd[0]


age_coef_dict = {
    18: 0.15190790211711158,
    19: 0.2352803285733691,
    20: 0.28917488419300963,
    21: 0.3235238459215871,
    22: 0.38186837776123933,
    23: 0.4656984696436133,
    24: 0.5180297776234675,
    25: 0.6476242253202377,
    26: 0.6864301922600129,
    27: 0.7556463866672888,
    28: 0.7403816046083588,
    29: 0.8065922023844624,
    30: 0.8232749662653331,
    31: 0.9226411592664203,
    32: 0.9170315557005079,
    33: 0.9228596827924056,
    34: 0.9451169052832582,
    35: 1.0359230366181207,
    36: 1.063442942909901,
    37: 1.079785503617365,
    38: 1.042574805277224,
    39: 1.0970818203288792,
    40: 1.1360596070532032,
    41: 1.2049768350353642,
    42: 1.1134926403507501,
    43: 1.3008241588488905,
    44: 1.163347869035451,
    45: 1.2355576326889808,
    46: 1.1751529816325876,
    47: 1.2238891949204627,
    48: 1.073537042852539,
    49: 1.1786549812413019,
    50: 1.2763276559659535,
    51: 1.183640097990281,
    52: 1.2539649986073098,
    53: 1.2621347169630002,
    54: 1.2090872328811855,
    55: 1.2075616293941958,
    56: 1.2005124884333722,
    57: 1.2205318363303777,
    58: 1.142836491755419,
    59: 1.2281186170170353,
    60: 1.1463919116971477,
    61: 1.2119927054379072,
    62: 1.2124939506093775,
    63: 1.2057006649277273,
    64: 1.1496953813548059,
    65: 1.1624445655681364,
    66: 1.2385043419939938,
    67: 1.4596112565399266,
    68: 1.3144227575971088,
    69: 1.2880937186120562,
    70: 1.1896895602046387,
    71: 1.198750867042334,
    72: 1.537697569224636,
    73: 1.4410072350681913,
    74: 1.098669435138111,
    75: 1.1062093558835275,
    76: 1.1062093558835275,
    77: 1.1062093558835275,
    78: 1.1062093558835275,
    79: 1.1062093558835275,
    80: 1.1062093558835275,
    81: 1.1062093558835275,
    82: 1.1062093558835275,
    83: 1.1062093558835275,
    84: 1.1062093558835275,
    85: 1.1062093558835275,
    86: 1.1062093558835275,
    87: 1.1062093558835275,
    88: 1.1062093558835275,
    89: 1.1062093558835275,
    90: 1.1062093558835275,
    91: 1.1062093558835275,
    92: 1.1062093558835275,
    93: 1.1062093558835275,
    94: 1.1062093558835275,
    95: 1.1062093558835275,
    96: 1.1062093558835275,
    97: 1.1062093558835275,
    98: 1.1062093558835275,
    99: 1.1062093558835275,
    100: 1.1062093558835275
}
